sstables/size_tiered_compaction_strategy.hh

```diff
@@ -302,6 +309,16 @@ size_tiered_compaction_strategy::get_sstables_for_compaction(column_family& cfs,
         return sstables::compaction_descriptor(std::move(most_interesting));
     }
 
+    if (sorted_sstable_pairs.size() > 1) {
+        // There is really nothing else to do within a tier, meaning we don't have a single tier with at least two
+        // SSTables. So we start going cross-tier. We start with the smallest SSTables possible, in the hopes that we
+        // will not have to fire a many-TB compaction for no reason, and something better will appear for us to do in
+        // the mean time.
+        std::vector<sstables::shared_sstable> cross_tier({sorted_sstable_pairs[0].first, sorted_sstable_pairs[1].first});
```

Comment:
Suppose we have a 1TB sstable and a 4TB sstable, and a trickle write load. Each flush generates a 1GB sstable.

The sstable size evolution (sizes in GB) will be:

1000 4000
1 1000 4000
1001 4000
1 1001 4000
1002 4000
1 1002 4000
1003 4000
...

Each such compaction has write amplification of ~1000:1. This will continue until we reach the size where the compacted sstable is considered in the same bucket as the 4TB one, when they will then be compacted. We will compact several petabytes until we reach that, probably killing the disk in the process.

Reply:
Note that this happens whatever we choose if the 4TB sstable doesn't exist. So we need to think again about idle compactions, this huge write amplification has a cost.

Reply:
That's why my original idea was to just force something like a major compaction at this point - essentially compacting everything together. If we were to do that, the 4TB SSTable would be compacted together with the rest of the data the first time around.

Reply:
Doesn't help, you get write amplification of 4000 again.

Reply:
Do you have any idea? I wasn't considering this to be too much of a problem because in order to compact another idle SSTable, the current idle has to finish first. Compacting many-TB can take quite some time, so the trickle load has to be slow to the point of creating one SSTable per hour for this problem to exist.

The one solution I can think of right of the bat is restricting it to one idle compaction per table per 24h period. So if we had idle-compacted a table (and in that case I'd strongly favor a major compaction, and not a closest-size), we start a timer to clear a flag and will only do another idle compaction once the flag is cleared.

Would that work ?

Reply:
This pile of heuristic is sure to fail in some way, let's try to understand the problem better instead.

Reply:
The problem seems to be that if we keep executing idle compactions, we'll wear out the disks too much because each compaction will have a very high amplification. A high amplification will have a high backlog, so we could start with the tables with the lowest backlog on idle, but that will still be an issue where we have a single table.

In your example, we could try to change the SSTable picking routine to pick 1000GB and 4000GB first, compact them, essentially just increasing the tolerante for things to be in the same tier.

The 1GB tables would be compacted together in their own tier for the foreseeable future, but is that a problem?

For the backlog, the calculation for one table is `Bi = Ei * log4 (T / Si)`, so as long as `T` is close to `Si`  the backlog will be low. So instead of considering things as part of the same tier with 50% similarity, we could increase that to higher numbers, and establish a maximum (95%?). This would keep the 1GB SSTables away from the pile.

However there is still the issue that we want to automatically be getting rid of tombstones for workloads that delete a lot, and that is essentially only done with a major compaction.

Right now we have an heuristic to compact an SSTable with itself if it has enough expirable tombstones. We could essentially get rid of that, and turn that one into a major compaction: if a single big SSTable has plenty of tombtones that are expirable and we're idle, we can then use that as a trigger of a full major compaction.

Does that sound better?

Reply:
It's still a bunch of heuristics. If tombstone compaction is only triggered on idle, then if we're never idle (perhaps due to activity in another table) we'll never compact those tombstones.

Each compaction has a cost and a benefit. For idle compactions, the cost is the amount of work involved (not critical since we're idle), disk wear, and less ability to react to new load (unless, perhaps, we stop idle compactions when the need arises). The benefit is a small reduction in read and space amplification. If the sstable size ratio is 100:1, then the reduction is just 1%.

"moving the backlog towards zero" is an artificial goal. Perhaps if we defined the backlog in a different way, that allows for one sstable per tier, we'd reach it without cross-tier compactions. Perhaps we should also consider how susceptible the disk is to wear (phase change disks aren't), or limit idle compaction bandwidth to something the disk can sustain (common is 3-5 disk writes per day, so for a 60TB disk that's 2.7GB/sec (not bad)).

Reply:
The reduction in space amplification can be much higher than 100:1, in fact it can get to 100% in the case where a last tier SSTable is comprised fully of expired tombstones. Note that I am not disputing what you said, just saying that this assumes an overwrite scenario with no deletions (in the other extreme, the no overwrite case the reduction is 0%)

Moving towards zero backlog may be an artificial goal - in which case we can definitely consider changing it, but moving SSTables to the last tier is not: as I said above, workloads with heavy deletion need that to get rid of tombstones.

Users run major compactions for that today, if we want to move to fully automate this, at some point we'll have to trigger something that resembles a major compaction. That is the end point of SizeTiered. The backlog is calculated to be zero at the point in which we have a single SSTable because of that, not arbitrarily.

Limiting bandwidth has a big problem in my view: the user will be better off running a major compaction manually, since it will get more bandwidth and finish faster. Finishing faster may also allow for other compactions to start running in other tables when they accumulate. In that sense, if we are to limit anything, I think my idea of limiting this to x times a day it is better: it is equivalent to bandwidth limit in the sense that over a period of time it will spend at most X MB/s on it, but it will potentially finish faster.

Reply:
If a compaction is used to reduce space amplification due to tombstones, then it is no longer an idle compaction. An idle compaction is something that can be deferred indefinitely (perhaps due to activity on another table). If we want to take that into account, it needs to go into the regular backlog calculations. Idle compaction is not a way to expire tombstones.

I do like moving sstables to the last tier. It is nice in that it improves layout, and our backlog calculation does describe the amount of work needed (similarly we should flush idle memtables). But it generates huge write amplification and we need to take that into account too.

Perhaps the time that a byte spends in a non-last tier should be taken into account. A byte that was just deposited into the first tier costs very little in terms of read amplification. A byte in the first tier for a year has costed us much more. But we also need to reconcile that with the current backlog estimation which I like and think is correct.

Reply:
I feel that you are right about maybe ditching the whole idle concept. Those compactions have to happen, they just have less priority than the others, like compactions in lower tiers.

I don't quite like the heuristic we have now about tombstones: it runs when nothing else is runnable (just local to the table, without the idle concept), and compacts a single SSTable. I think that if we do the work to rewrite an SSTable anyway, why not toss in SSTables from lower levels in the mix?

Perhaps we need to keep the backlog calculation as is, but take SSTable write times into account when deciding what to compact, similar to what a deadline scheduler would do.

So the backlog is still the same, but the priority of the highest level increases, say, logarithmically, with the creation date of the files.